### PR TITLE
Remove unsafe code in Uri ParseScheme

### PR DESCRIPTION
**Pull request by MihaZupan**
Instead of pinning and using `char*`, use the `string` directly. JIT is smart enough to eliminate bound checks in the first few loops, only needs `(uint)` help with the last few.

```c#
[Benchmark]
public Uri Http() => new Uri("http://contoso.com/");

[Benchmark]
public Uri Https() => new Uri("https://contoso.com/");

[Benchmark]
public Uri Wss() => new Uri("wss://contoso.com/");
```

Less unsafe => more perf :)

| Method |              Toolchain |     Mean |   Error |  StdDev |   Median | Ratio | RatioSD |
|------- |----------------------- |---------:|--------:|--------:|---------:|------:|--------:|
|   Http | \perf-base\CoreRun.exe | 207.8 ns | 1.63 ns | 8.32 ns | 212.8 ns |  1.10 |    0.05 |
|   Http |  \perf-new\CoreRun.exe | 189.2 ns | 0.95 ns | 4.66 ns | 186.9 ns |  1.00 |    0.00 |
|        |                        |          |         |         |          |       |         |
|  Https | \perf-base\CoreRun.exe | 208.8 ns | 0.86 ns | 4.37 ns | 211.3 ns |  1.05 |    0.02 |
|  Https |  \perf-new\CoreRun.exe | 198.1 ns | 0.88 ns | 4.39 ns | 195.2 ns |  1.00 |    0.00 |
|        |                        |          |         |         |          |       |         |
|    Wss | \perf-base\CoreRun.exe | 203.2 ns | 0.57 ns | 2.90 ns | 204.1 ns |  1.06 |    0.04 |
|    Wss |  \perf-new\CoreRun.exe | 191.8 ns | 1.21 ns | 6.00 ns | 195.6 ns |  1.00 |    0.00 |

### CODE REVIEW

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 1]
```diff
@@ -3530,25 +3525,13 @@ private unsafe void ParseRemaining()
...
-                if (end != length && end >= idx + 2 &&
-                    CheckKnownSchemes((long*)(uriString + idx), end - idx, ref syntax))
-                {
-                    return end + 1;
-                }
-            }
-
             //NB: A string must have at least 3 characters and at least 1 before ':'
-            if (idx + 2 >= length || end == idx)
+            if ((uint)(idx + 2) >= (uint)uriString.Length || end == idx)
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:3534]: Is the JIT able to eliminate bounds checks like `uriString[idx + 1]` based on this uint casting?  I thought that was beyond its ability today.

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 2]
```diff
@@ -3530,25 +3525,13 @@ private unsafe void ParseRemaining()
...
-            //
-            if (IntPtr.Size == 4)
-            {
-                // long = 4chars: The minimal size of a known scheme is 2 + ':'
-                if (end != length && end >= idx + 2 &&
-                    CheckKnownSchemes((long*)(uriString + idx), end - idx, ref syntax))
-                {
-                    return end + 1;
-                }
-            }
```
Comment (stephentoub) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:3548]: Why was this removed?  Did you benchmark on 32-bit as well and determine this wasn't helping?

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 1 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:3534]: Ah, you are correct, it will eliminate it for `idx + 2` but not for `idx` or `idx + 1`.
I will try to eliminate these as well!

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:3550]: I removed it because its behavior is wrong on big-endian.
I ran the benchmark for 32bit and there is a ~4% regression in the worst-case.

If we believe this kind of optimization is necessary, I can add an alternative version of it - initial testing showed it can save ~15% from ctor time for simple Uris.

Comment (stephentoub) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:3550]: > I removed it because its behavior is wrong on big-endian. I ran the benchmark for 32bit and there is a ~4% regression in the worst-case.

Rather than deleting it, can you instead just make it:
```C#
if (IntPtr.Size == 4 && BitConverter.IsLittleEndian)
```
?

**Comment by stephentoub** [comment]
@MihaZupan, are you still working on this?  Thanks.

**Comment by stephentoub** [comment]
@MihaZupan, do you plan to finish this soon?  It's been open for 5 months and untouched since June.
